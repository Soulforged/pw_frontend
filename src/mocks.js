//@flow

//FIXME remove all mocks from here when they're no longer necessary
/* eslint-disable global-require,import/prefer-default-export */
export default {
  summary: require("./data/summary.json"),
  trendsByDate: require("./data/trendsByDate.json"),
  trendsByInstitution: require("./data/trendsByInstitution.json"),
  errors: require("./data/errors.json"),
  users: require("./data/users.json"),
  login: require("./data/login.json"),
  user: require("./data/user.json"),
  userCreate: require("./data/userCreate.json"),
  role: require("./data/role.json")
};

/* eslint-disable */
