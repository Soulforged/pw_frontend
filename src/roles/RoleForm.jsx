//@flow
import React from "react";
import { omit } from "lodash";
import { Form, Text, Select, RadioGroup, Radio } from "react-form";
import { branch, compose, renderComponent } from "recompose";
import { boundLifecycle } from "src/recompose-ext";
import { Loading } from "src/components";

type Props = {
  item: Object,
  // saveRole: (Object) => void,
  saving: boolean,
  edit: boolean,
  closeNew: (Event) => void,
};

const bunitOptions = [
  { label: "Finance", value: 6 },
  { label: "Accounts", value: 117 },
];

const statusOptions = [
  { label: "Active", value: 1 },
  { label: "Inactive", value: 0 },
];

const ButtonOrLoading = ({ saving }: { saving: boolean }) => (
  saving ? <Loading loading /> : <button type="submit" className="btn btn-default">SAVE</button>
);

type Item = {
  item: Object
};

const StatusField = ({ item }: Item) => (
  <div className="row modal-p">
    <div className="col-md-6 col-sm-6">
      <label htmlFor="status">Status:</label>
    </div>
    <div className="col-md-6 col-sm-6">
      <Select
        id="status"
        className="form-control"
        field="status"
        options={statusOptions}
        value={item.status}
      />
    </div>
  </div>
);

const BusinessUnitSelector = ({ item }: Item) => (
  <div className="row modal-p">
    <div className="col-md-6 col-sm-6">
      <label htmlFor="businessUnitName">Business Unit:</label>
    </div>
    <div className="col-md-6 col-sm-6">
      <Select
        id="businessUnitName"
        className="form-control"
        field="businessUnitId"
        options={bunitOptions}
        value={item.businessUnitId}
        required
      />
    </div>
  </div>
);

const preValidate = values => (
  values.checked ? values : omit(values, "businessUnitId")
);

const validate = ({ checked, businessUnitId }) => (
  {
    businessUnitId: { error: checked && !businessUnitId && "You must specify" }
  }
);

const Component = (props: Props) => {
  const {
    item,
    // saveRole,
    saving,
    edit = item.id,
    closeNew,
  } = props;
  return (
    <div>
      <div id="main-pnl" className="pnls">
        <h4 className="form-header trebuchet text-center bold">
          {edit ? `Edit Role ${item.id}` : "New Role"}
          <button className="add-new pointer pull-right bold" onClick={closeNew}>
            <i className="fa fa-close theme" />
          </button>
        </h4>
        <div className="add-pnl-cnt">
          <Form
            onSubmit={submittedValues => console.log(submittedValues)}
            defaultValues={{ ...item, checked: item.id ? "1" : "" }}
            preValidate={preValidate}
            validate={validate}
            validateOnSubmit
          >
            {formApi => (
              <form id="role_frm" onSubmit={formApi.submitForm}>
                <div className="row modal-p">
                  <div className="col-md-6 col-sm-6">
                    <label htmlFor="roleName">Role Name:</label>
                  </div>
                  <div className="col-md-6 col-sm-6">
                    <Text
                      field="name"
                      id="roleName"
                      className="form-control"
                      placeholder="Role name"
                      required
                    />
                  </div>
                </div>
                <div className="row modal-p">
                  <div className="col-md-6 col-sm-6">
                    <label htmlFor="roleType">Role Type:</label>
                  </div>
                  <div className="col-md-6 col-sm-6">
                    <RadioGroup field="checked">
                      <label htmlFor="radioCompany">Company Wide</label>
                      <Radio id="radioCompany" value="" />
                      <label htmlFor="radioBu">Business Unit Specific</label>
                      <Radio id="radioBu" value="1" />
                    </RadioGroup>
                  </div>
                </div>
                {formApi.values.checked && <BusinessUnitSelector item={item} />}
                {edit && <StatusField item={item} />}
                <hr />
                <div>
                  {formApi.errors
                    && Object.keys(formApi.errors).map(e => <div key={e}>{formApi.errors[e]}</div>)}
                </div>
                <article className="modal-p text-center">
                  <ButtonOrLoading saving={saving} />
                </article>
              </form>
            )}
          </Form>
        </div>
      </div>
    </div>
  );
};

const SpecLoading = () => <Loading loading />;
const NoData = () => <div id="main-pnl" className="pnls">Role(s) not found</div>;

export default compose(
  boundLifecycle({
    didMount: ({ id, fetchRole, item }) => !item && fetchRole(id)
  }),
  branch(({ fetching }) => fetching, renderComponent(SpecLoading)),
  branch(({ item }) => item == null, renderComponent(NoData))
)(Component);
