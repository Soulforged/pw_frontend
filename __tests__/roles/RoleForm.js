//@flow
import React from "react";
import RoleForm from "src/roles/RoleForm";
import renderer from "react-test-renderer";
import mocks from "src/mocks";

it("matches snapshot while fetching", () => {
  const rendered = renderer.create(<RoleForm fetchRole={jest.fn()} fetching />).toJSON();
  expect(rendered).toMatchSnapshot();
});

it("matches snapshot no data", () => {
  const rendered = renderer.create(<RoleForm fetchRole={jest.fn()} />).toJSON();
  expect(rendered).toMatchSnapshot();
});

it("matches snapshot while saving", () => {
  const rendered = renderer.create((
    <RoleForm fetchRole={jest.fn()} saving item={mocks.role} />
  )).toJSON();
  expect(rendered).toMatchSnapshot();
});

it("matches snapshot while not saving", () => {
  const rendered = renderer.create(<RoleForm fetchRole={jest.fn()} item={mocks.role} />).toJSON();
  expect(rendered).toMatchSnapshot();
});
