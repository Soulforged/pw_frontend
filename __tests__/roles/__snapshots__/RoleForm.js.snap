// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot no data 1`] = `
<div
  className="pnls"
  id="main-pnl"
>
  Role(s) not found
</div>
`;

exports[`matches snapshot while fetching 1`] = `
<div
  id="loader"
  style={Object {}}
>
  <img
    alt="loading something..."
    className="loader"
    src="loader.gif"
  />
</div>
`;

exports[`matches snapshot while not saving 1`] = `
<div
  className="pnls"
  id="main-pnl"
>
  <h4
    className="form-header trebuchet text-center bold"
  >
    Edit Role 1
    <button
      className="add-new pointer pull-right bold"
      onClick={undefined}
    >
      <i
        className="fa fa-close theme"
      />
    </button>
  </h4>
  <div
    className="add-pnl-cnt"
  >
    <form
      id="role_frm"
      onSubmit={[Function]}
    >
      <div
        className="row modal-p"
      >
        <div
          className="col-md-6 col-sm-6"
        >
          <label
            htmlFor="roleName"
          >
            Role Name:
          </label>
        </div>
        <div
          className="col-md-6 col-sm-6"
        >
          <input
            className="form-control"
            id="roleName"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Role name"
            required={true}
            value="role"
          />
        </div>
      </div>
      <div
        className="row modal-p"
      >
        <div
          className="col-md-6 col-sm-6"
        >
          <label
            htmlFor="roleType"
          >
            Role Type:
          </label>
        </div>
        <div
          className="col-md-6 col-sm-6"
        >
          <div>
            <input
              checked={true}
              id="radioCompany"
              onBlur={[Function]}
              onChange={[Function]}
              type="radio"
              value=""
            />
            <label
              htmlFor="radioCompany"
            >
              Company Wide
            </label>
            <input
              checked={false}
              id="radioBu"
              onBlur={[Function]}
              onChange={[Function]}
              type="radio"
              value="1"
            />
            <label
              htmlFor="radioBu"
            >
              Business Unit Specific
            </label>
          </div>
        </div>
      </div>
      
      <div
        className="row modal-p"
      >
        <div
          className="col-md-6 col-sm-6"
        >
          <label
            htmlFor="status"
          >
            Status:
          </label>
        </div>
        <div
          className="col-md-6 col-sm-6"
        >
          <select
            className="form-control"
            id="status"
            onBlur={[Function]}
            onChange={[Function]}
            value={1}
          >
            <option
              disabled={true}
              value={0}
            >
              Select One...
            </option>
            <option
              disabled={undefined}
              value={1}
            >
              Active
            </option>
            <option
              disabled={undefined}
              value={2}
            >
              Inactive
            </option>
          </select>
        </div>
      </div>
      <hr />
      <div />
      <article
        className="modal-p text-center"
      >
        <button
          className="btn btn-default"
          type="submit"
        >
          SAVE
        </button>
      </article>
    </form>
  </div>
</div>
`;

exports[`matches snapshot while saving 1`] = `
<div
  className="pnls"
  id="main-pnl"
>
  <h4
    className="form-header trebuchet text-center bold"
  >
    Edit Role 1
    <button
      className="add-new pointer pull-right bold"
      onClick={undefined}
    >
      <i
        className="fa fa-close theme"
      />
    </button>
  </h4>
  <div
    className="add-pnl-cnt"
  >
    <form
      id="role_frm"
      onSubmit={[Function]}
    >
      <div
        className="row modal-p"
      >
        <div
          className="col-md-6 col-sm-6"
        >
          <label
            htmlFor="roleName"
          >
            Role Name:
          </label>
        </div>
        <div
          className="col-md-6 col-sm-6"
        >
          <input
            className="form-control"
            id="roleName"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Role name"
            required={true}
            value="role"
          />
        </div>
      </div>
      <div
        className="row modal-p"
      >
        <div
          className="col-md-6 col-sm-6"
        >
          <label
            htmlFor="roleType"
          >
            Role Type:
          </label>
        </div>
        <div
          className="col-md-6 col-sm-6"
        >
          <div>
            <input
              checked={true}
              id="radioCompany"
              onBlur={[Function]}
              onChange={[Function]}
              type="radio"
              value=""
            />
            <label
              htmlFor="radioCompany"
            >
              Company Wide
            </label>
            <input
              checked={false}
              id="radioBu"
              onBlur={[Function]}
              onChange={[Function]}
              type="radio"
              value="1"
            />
            <label
              htmlFor="radioBu"
            >
              Business Unit Specific
            </label>
          </div>
        </div>
      </div>
      
      <div
        className="row modal-p"
      >
        <div
          className="col-md-6 col-sm-6"
        >
          <label
            htmlFor="status"
          >
            Status:
          </label>
        </div>
        <div
          className="col-md-6 col-sm-6"
        >
          <select
            className="form-control"
            id="status"
            onBlur={[Function]}
            onChange={[Function]}
            value={1}
          >
            <option
              disabled={true}
              value={0}
            >
              Select One...
            </option>
            <option
              disabled={undefined}
              value={1}
            >
              Active
            </option>
            <option
              disabled={undefined}
              value={2}
            >
              Inactive
            </option>
          </select>
        </div>
      </div>
      <hr />
      <div />
      <article
        className="modal-p text-center"
      >
        <div
          id="loader"
          style={Object {}}
        >
          <img
            alt="loading something..."
            className="loader"
            src="loader.gif"
          />
        </div>
      </article>
    </form>
  </div>
</div>
`;
